proposals/0016-simd-application-fees.md

```diff
@@ -141,295 +125,367 @@ application is not CPI'd to. There were multiple possible approaches discussed
+  executed and the payer will be reimbursed the overpaid amount.
+
+This instruction cannot be CPI'ed into.
+
+### UpdateApplicationFees
+
+This mechanism will set the application fees for an account in the `Account`
+structure. The account must already exist and should be rent-free to change its
+application fees. Only the account authority must be allowed to update the
+application fees of an account. The application fee set will permanently update
+the account in the ledger. This mechanism can be reused with fees set to 0 to
+remove application fees on an account.
+
+### Rebate
+
+This mechanism should be called by the authority program to issue a rebate.
+Rebate takes amount of lamports to be rebated as input. In case of multiple
+rebates by the program during the execution, only the highest amount of rebate
+will be taken into account. The rebated amount is always the minimum of rebate
+issued by the program and the application fees on the account. If program
+rebates `U64::MAX` it means all the application fees on the account are rebated.
+Rebate amount cannot be 0.
+
+### Looking at common cases
+
+##### No application fees involved
+
+* A payer does not include `PayApplicationFees` in the transaction. The
+  transaction does not write lock any accounts with application fees. Then the
+  transaction is executed without the application fee feature. The payer ends
+  up paying other fees.
+
+* A payer includes `PayApplicationFees(app fees)` in the transaction but none of
+  the accounts have any application fees. This case is considered as overpay case.
+  The payer balance is checked for `other fees + app fees`.
+  - The payer does not have enough balance: Transaction fails with an error
+    `Insufficient Balance` and the transaction is not even scheduled for
+    execution.
+  - The payer has enough balance then the transaction is executed and application
+    fees paid are transferred back to the payer in any case.
+  
+  Note in this case
+    even if there are no application fees involved the payer balance is checked
+    for application fees.
+
+##### Application fees are involved
+
+* Fees not paid case:
+
+  A payer does not include `PayApplicationFees` in the transaction. The
+  transaction includes one or more accounts with application fees. Then the
+  transaction is failed with an error `ApplicationFeesNotPaid`. The program is
+  not executed at all. The payer ends up paying only other fees.
+
+* Fees paid no rebates case:
+
+  A payer includes instruction `PayApplicationFees(200)` in the transaction.
+  There are accounts (`accA`, `accB`) that are write-locked by the transaction
+  and each of them has an application fees of `100` lamports. Consider that the
+  program does not have any rebate mechanism. Then in any case (execution fails
+  or succeeds) `accA` will receive `100` lamports, `accB` will receive `100`
+  lamports. The payer will end up paying `other fees` + `200` lamports.
+
+* Fees paid full rebates case:
```

Comment:
full rebate is not so relevant, would also remove

Reply:
I think it is better to give at least one example of full rebate right?

Reply:
Kind of how everything should work by default